### ISSUE TITLE
typedef for arrow functions demanding function-specific return

### ISSUE THREAD

**Issue by thekevinscott** [palantir/tslint#4462]
Hello,

I'm confused by `typedef` behavior, but not sure it's a bug (it may be mistaken understanding on my part).

I'm trying to use the `typedef` rule to enforce arrow functions to have an explicit return type. This is working, but it appears that those return types must be defined at the function itself, instead of in some interface.

[Github repo is here.](https://github.com/thekevinscott/test-typedef)

Is this a bug? Or is this behavior intended?

### Bug Report

- __TSLint version__: 5.12.1
- __TypeScript version__: 3.2.2
- __Running TSLint via__: (pick one) CLI

#### TypeScript code being linted

```ts
interface IFoo {
  (): boolean;
}

// this code throws a linting issue
export const notWorking: IFoo = () => true;

// this code works fine, though the boolean return type seems redundant
export const working: IFoo = (): boolean => true;
```

```json
{
  "rules": {
    "typedef": [
      true,
      "arrow-call-signature"
    ]
  }
}
```

#### Actual behavior
The `notWorking` function fails the linting rule for not providing a call signature.

#### Expected behavior
I would expect that the `notWorking` function would not fail the linting rule, since the `IFoo` interface defines a return type.

**Comment by JoshuaKGoldberg** [comment]
Agreed, I think the `typedef` rule should be relaxed for functions (arrow and others) who are assigned a type in a parent variable declaration. Especially given how much TypeScript's type inference has improved since the rule was first added...

Sorry for the delay @thekevinscott! Accepting PRs to change the rule.

**Comment by VincentLanglet** [comment]
It works fine for me. I understand the type definitions like a _type documentation_.
And then, I expect the function to be _documented_. This is not related to inference/being redundant.

And for example, what if I have
```
interface IFoo {
  (): boolean | string;
}

export const working: IFoo = (): boolean => true;
```

That's not the same than
```
export const working: IFoo = () => true;
```
Cause someone can change this code this way, without any error
```
export const working: IFoo = () => 'true';
```

**Comment by adidahiya** [comment]
I think there are users on both sides of this "`typedef` as documentation" / "`typedef` as catching things just outside the scope of tsc's inference capabilities" debate... I'm open to supporting both use cases as long as there is a reasonable proposal for how to create a rule options API which is not overlay complex, but I don't have a good idea for how to do that right now

**Comment by JoshuaKGoldberg** [comment]
☠️ *TSLint's time has come!* ☠️ 

TSLint is no longer accepting most feature requests per #4534. See **[typescript-eslint.io](https://typescript-eslint.io)** for the new, shiny way to lint your TypeScript code with **ESLint**. ✨ 

It was a pleasure open sourcing with you all!

**Comment by JoshuaKGoldberg** [comment]
🤖 Beep boop! 👉  **TSLint is deprecated** 👈 _(#4534)_  and you should **switch to [typescript-eslint](https://typescript-eslint.io)**! 🤖 

🔒 This issue is being locked to prevent further unnecessary discussions. Thank you! 👋
